src/ufuncs.c

```diff
@@ -1006,39 +1096,39 @@ static void bounds_func(char **args, npy_intp *dimensions,
             *x1 = *y1 = *x2 = *y2 = NPY_NAN;
         } else {
             /* construct the envelope */
-            envelope = GEOSEnvelope_r(context, in1);
-            if (envelope == NULL) { return; }
-            size = GEOSGetNumCoordinates_r(context, envelope);
+            envelope = GEOSEnvelope_r(ctx, in1);
+            if (envelope == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+            size = GEOSGetNumCoordinates_r(ctx, envelope);
 
             /* get the bbox depending on the number of coordinates in the envelope */
             if (size == 0) {  /* Envelope is empty */
                 *x1 = *y1 = *x2 = *y2 = NPY_NAN;
             } else if (size == 1) {  /* Envelope is a point */
-                if (!GEOSGeomGetX_r(context, envelope, x1)) { goto fail; }
-                if (!GEOSGeomGetY_r(context, envelope, y1)) { goto fail; }
+                if (!GEOSGeomGetX_r(ctx, envelope, x1)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+                if (!GEOSGeomGetY_r(ctx, envelope, y1)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
                 *x2 = *x1;
                 *y2 = *y1;
             } else if (size == 5) {  /* Envelope is a box */
-                ring = GEOSGetExteriorRing_r(context, envelope);
-                if (ring == NULL) { goto fail; }
-                coord_seq = GEOSGeom_getCoordSeq_r(context, ring);
-                if (coord_seq == NULL) { goto fail; }
-                if (!GEOSCoordSeq_getX_r(context, coord_seq, 0, x1)) { goto fail; }
-                if (!GEOSCoordSeq_getY_r(context, coord_seq, 0, y1)) { goto fail; }
-                if (!GEOSCoordSeq_getX_r(context, coord_seq, 2, x2)) { goto fail; }
-                if (!GEOSCoordSeq_getY_r(context, coord_seq, 2, y2)) { goto fail; }
+                ring = GEOSGetExteriorRing_r(ctx, envelope);
+                if (ring == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+                coord_seq = GEOSGeom_getCoordSeq_r(ctx, ring);
+                if (coord_seq == NULL) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+                if (!GEOSCoordSeq_getX_r(ctx, coord_seq, 0, x1)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+                if (!GEOSCoordSeq_getY_r(ctx, coord_seq, 0, y1)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+                if (!GEOSCoordSeq_getX_r(ctx, coord_seq, 2, x2)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
+                if (!GEOSCoordSeq_getY_r(ctx, coord_seq, 2, y2)) { errstate = PGERR_GEOS_EXCEPTION; goto finish; }
             } else {
-                PyErr_Format(PyExc_ValueError, "Could not determine bounds from an envelope with %d coordinate pairs", size);
-                goto fail;
+                errstate = PGERR_GEOMETRY_TYPE;
```

Comment:
Can be done as a follow-up if we would want this, but I am wondering: would it be possible to set a custom error message as well with the new error handling machinery? 
Since before, in the above case a more specific (informative) error message was raised.

Reply:
I think that is possible, although I am not sure if it is necessary. I'd like to keep it out of scope now.